ayeletK/assignment1/Q1_solution.cpp

```diff
@@ -0,0 +1,114 @@
+#include <string>
+#include <vector>
+#include <cctype>
+#include <iostream>
+#include <algorithm>
+
+using namespace std;
+
+/**
+    Determine if a character is a punctuation mark
+	  from the next list: ‘ ‘/’,’/’.’/’!’/’?’
+    @param c - the character to be checked
+    @return true if the character is a punctuation mark,
+	    false otherwise.
+    
+    example:
+    c = "!"
+    ------------------
+    true
+*/
+bool is_punctuation_mark(char c){
+    if (c == ' ' || c == ',' || c == '.' ||
+        c == '!' || c == '?'){
+            return true;
+        }
+    return false;
+}
+
+/**
+    Split a sentence by white spaces and punctuation marks
+    [',','.','!','?'] into a vector and sort each word 
+    ("aba" --> "aab")
+
+    @param str - the sentence to be splitted
+    @param case_sensitive - a flag determins if we are 
+      working in a case sensitive mode or not.
+      if not 'case_sensitive' then all letters will be 
+      lowered to lower case.
+    @return a vector of strings, each string in the result
+      vector represent a sorted word from the original
+      sentence.
+    
+    example:
+    str = "Like a G6"
+    case_sensitive = false
+    ------------------
+    words: <"eikl", "a", "6g">
+*/
+vector<string> split_to_words(string str, bool case_sensitive){
+    vector<string> words;
+    string curr_str = "";
+    for (int i = 0; i<str.length() ; ++i){
+    // add new word to words vector when reaching to a punctuation mark
+		if (is_punctuation_mark(str[i]) && !curr_str.empty()){
+		    sort(curr_str.begin(), curr_str.end());
+			words.push_back(curr_str);
+			curr_str.clear();
+			while (i+1<str.length() && is_punctuation_mark(str[i+1])){
+			    ++i;
+			}
+		}
+    // add character to the current word
+		else {
+		    char curr_char = str[i];
+		    if (!case_sensitive){
+		        curr_char = tolower(curr_char);
+		    }
+		    curr_str += curr_char;
+		}
+	}
+  // save the last word in the sentence, in case it ends with a word
+	if (!curr_str.empty()){
+	    sort(curr_str.begin(), curr_str.end());
```

Comment:
This 'sort and push' is a code duplication, it appears similarly in lines 55-56.
Such duplications are best avoided; for example, when someone edits this code (let's say, replaces the sort with a different, more efficient function) they can change one occurrence of this and forget to change the other one.

A solution to this is to move the sort to the end of the function, and sort the vector after you are done adding all the words to it.
Another solution is take the sort out of this function completely as I suggested above.

Reply:
Note taken. Do you think I can avoid checking the case the sentence ended with a word outside the "for" loop?